### PR TITLE
Add (unused) HIP API to the Context object.

### PR DESCRIPTION
**Pull request by ezyang**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #14414 Delete at::current_device(), Context::current_device() and Context::getNumGPUs()&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13218540/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #14499 Delete hasCuDNN from Context.&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13241355/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #14500 Replace THCState_getCurrentStream with direct at::cuda::getCurrentCUDAStream()&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13241401/)
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#14623 Add (unused) HIP API to the Context object.**&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13277246/)

This is the last piece we need before we can start doing out-of-place
HIPify on ATen.  These APIs are not actually used at the moment;
as we still do in-place HIPify which uses CUDA.

Differential Revision: [D13277246](https://our.internmc.facebook.com/intern/diff/D13277246/)

### CODE REVIEW

File: aten/src/ATen/detail/HIPHooksInterface.cpp [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+#include <memory>
+#include <mutex>
+
+namespace at {
+namespace detail {
+
+const HIPHooksInterface& getHIPHooks() {
+  static std::unique_ptr<HIPHooksInterface> hip_hooks;
+  // NB: The once_flag here implies that if you try to call any HIP
+  // functionality before libATen_cuda.so is loaded, HIP is permanently
```
Comment (gchanan) [inline aten/src/ATen/detail/HIPHooksInterface.cpp]: perhaps you should clarify the libATen_cuda.so situation in the comment (are we going to have a libHIP?)

File: aten/src/ATen/detail/HIPHooksInterface.cpp [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+  // disabled for that copy of ATen.  In principle, we can relax this
+  // restriction, but you might have to fix some code.  See getVariableHooks()
```
Comment (gchanan) [inline aten/src/ATen/detail/HIPHooksInterface.cpp]: you don't need to change this, but I bet you this comment lasts longer than getVariableHooks exists :).

**Review by gchanan — APPROVED** [review]

File: aten/src/ATen/detail/HIPHooksInterface.cpp (hunk 1 above)
Comment (ezyang) [reply aten/src/ATen/detail/HIPHooksInterface.cpp]: I just deleted this comment XD

Comment (ezyang) [reply aten/src/ATen/detail/HIPHooksInterface.cpp]: But yes, we'll have a libATen_hip once we get out of place hipify

**Pull request closed without merging**